A unit test is flaky: TestGracefulStopCommand

**What happened**:

test passes

**What you expected to happen**:

not passes

**How to reproduce it**:

```
--- FAIL: TestGracefulStopCommand (0.00s)
    --- FAIL: TestGracefulStopCommand/graceful_stop_after_very_short_time (0.00s)
        binary_test.go:48: 
            	Error Trace:	/home/runner/work/pipecd/pipecd/pkg/app/launcher/cmd/launcher/binary_test.go:48
            	Error:      	Should be false
            	Test:       	TestGracefulStopCommand/graceful_stop_after_very_short_time
```

**Environment**:
- `piped` version:
- `control-plane` version:
- Others:

--- a/pkg/app/launcher/cmd/launcher/binary.go
+++ b/pkg/app/launcher/cmd/launcher/binary.go
@@ -21,6 +21,7 @@ import (
 	"os"
 	"os/exec"
 	"path/filepath"
+	"sync/atomic"
 	"syscall"
 	"time"
 
@@ -29,7 +30,8 @@ import (
 
 type command struct {
 	cmd       *exec.Cmd
-	stoppedCh chan error
+	stoppedCh chan struct{}
+	result    atomic.Pointer[error]
 }
 
 func (c *command) IsRunning() bool {
@@ -50,9 +52,16 @@ func (c *command) GracefulStop(period time.Duration) error {
 	select {
 	case <-timer.C:
 		c.cmd.Process.Kill()
-		return <-c.stoppedCh
-	case err := <-c.stoppedCh:
-		return err
+		<-c.stoppedCh
+		if perr := c.result.Load(); perr != nil {
+			return *perr
+		}
+		return nil
+	case <-c.stoppedCh:
+		if perr := c.result.Load(); perr != nil {
+			return *perr
+		}
+		return nil
 	}
 }
 
@@ -68,11 +77,12 @@ func runBinary(execPath string, args []string) (*command, error) {
 
 	c := &command{
 		cmd:       cmd,
-		stoppedCh: make(chan error, 1),
+		stoppedCh: make(chan struct{}),
+		result:    atomic.Pointer[error]{},
 	}
 	go func() {
 		err := cmd.Wait()
-		c.stoppedCh <- err
+		c.result.Store(&err)
 		close(c.stoppedCh)
 	}()
 

--- a/pkg/app/launcher/cmd/launcher/binary_test.go
+++ b/pkg/app/launcher/cmd/launcher/binary_test.go
@@ -15,6 +15,7 @@
 package launcher
 
 import (
+	"strconv"
 	"testing"
 	"time"
 
@@ -45,6 +46,37 @@ func TestGracefulStopCommand(t *testing.T) {
 
 			assert.True(t, cmd.IsRunning())
 			cmd.GracefulStop(tc.stopAfter)
+			assert.False(t, cmd.IsRunning())
+		})
+	}
+}
+
+func TestGracefulStopCommandResult(t *testing.T) {
+	testcases := []struct {
+		name      string
+		exitCode  int
+		assertion assert.ErrorAssertionFunc
+	}{
+		{
+			name:      "successfully exit",
+			exitCode:  0,
+			assertion: assert.NoError,
+		},
+		{
+			name:      "exit with an error",
+			exitCode:  1,
+			assertion: assert.Error,
+		},
+	}
+
+	for _, tc := range testcases {
+		t.Run(tc.name, func(t *testing.T) {
+			cmd, err := runBinary("sh", []string{"-c", "exit " + strconv.Itoa(tc.exitCode)})
+			require.NoError(t, err)
+			require.NotNil(t, cmd)
+
+			time.Sleep(100 * time.Millisecond) // to avoid GracefulStop executed before the command exits
+			tc.assertion(t, cmd.GracefulStop(time.Second))
 			assert.False(t, cmd.IsRunning())
 		})
 	}

